e2e/framework.go

```diff
@@ -60,10 +61,37 @@ func initTracer(name string) *sdktrace.TracerProvider {
 }
 
 type cluster struct {
-	t      *testing.T
-	nodes  map[string]*node
-	tracer *sdktrace.TracerProvider
-	hook   transportHook
+	t               *testing.T
+	lock            sync.Mutex
+	nodes           map[string]*node
+	tracer          *sdktrace.TracerProvider
+	hook            transportHook
+	sealedProposals []*pbft.SealedProposal
+	lastProposer    pbft.NodeID
+}
+
+// getSyncIndex returns a index up to which the node is synced with the network
+func (c *cluster) getSyncIndex(node string) uint64 {
+	return c.nodes[node].getSyncIndex()
+}
+
+// insertFinalProposal inserts final proposal from the node to the cluster and returns index up to which the node is synced
+func (c *cluster) insertFinalProposal(p *pbft.SealedProposal) uint64 {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if len(c.sealedProposals) > 0 {
+		v := c.sealedProposals[len(c.sealedProposals)-1]
```

Comment:
Sorry, I know that I already made a comment here but I though about something different. Could there be a situation of many nodes in a fast pbft protocol in which some subset with quorum start another round and finish before another set finishes writing?

I think it is safer to assume that `p.Number` == your index in `c.sealedProposals` and do that check instead.

```
lastIndex := len(c.sealedProposals)
if lastIndex + 1 == p.Number {
   // new proposal
   c.sealedProposals = append(c.sealedProposals, p)
} else {
   // already in, validate
  if !p.Equal(c.sealedProposals[p.Number] {
     panic(...)
  }
}
```